icu4c/source/test/intltest/ustrtest.cpp

```diff
@@ -2310,3 +2311,21 @@ void UnicodeStringTest::TestUnicodeStringInsertAppendToSelf() {
     str.insert(2, sub);
     assertEquals("", u"abbcdcde", str);
 }
+
+void UnicodeStringTest::TestLargeAppend() {
+    IcuTestErrorCode status(*this, "TestLargeAppend");
+    // Make a large UnicodeString
+    int32_t len = 0xAFFFFFF;
+    UnicodeString str;
+    char16_t *buf = str.getBuffer(len);
+    for (int32_t i = 0; i < len; i++) {
+      *buf++ = u'a';
+    }
+    str.releaseBuffer(len);
+    UnicodeString dest;
+    // Append it 16 times
+    for (int32_t i = 0; i < 16; i++) {
+      dest.append(str);
+    }
+    // Pass the test if it does not crash and reach here.
```

Comment:
"On 32 bit platforms the test might fail in unexpected ways from memory allocation failures. Probably would cause the string to go bogus sooner than expected." 
how should I wrote that test then?